### PR TITLE
[Merged by Bors] - feat: updating type-guards to use Zod parsing for more reliable runtime validation (PL-753)

### PR DESCRIPTION
**Pull request by zhihil**
**Fixes or implements PL-753**

### Brief description. What is this change?

Updated the request type guards in `general-runtime` to use Zod parsing instead. This provides more reliable runtime validation. 

Previously, the `general-runtime` only checke the `type` property of the request object, which was unreliable as the actual data sent to the API may not match the type declaration. By using Zod, we provide complete validation that the incoming request object matches the type. 

### Implementation details. How do you make this change?

The main changes are found here: `lib/services/runtime/types.ts`. This PR replaces many of the simple type-guards with an `DTO.isXRequest()` function. Under the hood, this function uses Zod parsing, i.e, `XRequestDTO.parse(value).success`

Additionally, Zod parsing has been added to the following controllers which expect an object of type `RuntimeRequest` in the body:
1. `lib/controllers/interact.ts`
2. `lib/controllers/public/index.ts`
4. `lib/controllers/stateManagement/index.ts`

This ensures that the incoming request objects are being validated.

All other changes simply apply the new type-guards.

### Setup information

N/A

### Deployment Notes

N/A

### Related PRs

1. https://github.com/voiceflow/libs/pull/522
5. https://github.com/voiceflow/general-runtime/pull/790
6. https://github.com/voiceflow/creator-app/pull/8046

### Checklist

- [ ] Breaking changes have been communicated, including:
    - New required environment variables
    - Renaming of interfaces (API routes, request/response interface, etc)
- [ ] New environment variables have [been deployed](https://www.notion.so/voiceflow/Add-Environment-Variables-be1b0136479f45f1adece7995a7adbfb)
- [ ] Appropriate tests have been written
    - Bug fixes are accompanied by an updated or new test
    - New features are accompanied by a new test

### CODE REVIEW

File: lib/services/runtime/index.ts [hunk 1]
```diff
@@ -64,7 +64,7 @@ class RuntimeManager extends AbstractManager<{ utils: typeof utils }> implements
     }
 
     if (isPathRequest(request)) {
-      runtime.variables.set(VoiceflowConstants.BuiltInVariable.LAST_UTTERANCE, request.payload.label);
+      runtime.variables.set(VoiceflowConstants.BuiltInVariable.LAST_UTTERANCE, request.payload?.label);
```
Comment (zhihil) [inline lib/services/runtime/index.ts:67]: `.label` is already potentially undefined, so the optional chaining shouldn't cause a huge deviation from existing behaviour.

File: lib/services/aiAssist.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ class AIAssist extends AbstractManager implements ContextHandler {
     }
 
     if (isPathRequest(request)) {
-      return request.payload.label ?? null;
+      return request.payload?.label ?? null;
```
Comment (zhihil) [inline lib/services/aiAssist.ts:28]: `.label` here was already potentially `undefined`, so adding this optional chaining should not create big issues.

File: lib/services/runtime/types.ts [hunk 3]
```diff
@@ -13,32 +13,32 @@ export interface Prompt {
...
+export const isTextRequest = (request?: RuntimeRequest): request is BaseRequest.TextRequest =>
   !!request && BaseRequest.isTextRequest(request) && typeof request.payload === 'string';
 
 /**
  * Intent request is being reused for both Alexa events and intent events. To distinguish them we check
  * if `request.payload.data` exists, if so this is an Alexa event request
  * otherwise it is a normal intent request
  */
-export const isIntentRequest = (request?: RuntimeRequest | null): request is BaseRequest.IntentRequest =>
+export const isIntentRequest = (request?: RuntimeRequest): request is BaseRequest.IntentRequest =>
```
Comment (zhihil) [inline lib/services/runtime/types.ts:24]: The `| null` here is redundant because `RuntimeRequest` already comes with `| null`.

**Review by zhihil — COMMENTED** [review]
Context:

File: lib/services/runtime/types.ts [hunk 4]
```diff
@@ -13,34 +14,34 @@ export interface Prompt {
   voice?: string;
 }
 
-export const isTextRequest = (request?: RuntimeRequest | null): request is BaseRequest.TextRequest =>
-  !!request && BaseRequest.isTextRequest(request) && typeof request.payload === 'string';
+export const isTextRequest = (request?: RuntimeRequest): request is BaseRequest.TextRequest =>
+  !!request && isTextRequest(request) && typeof request.payload === 'string';
```
Comment (pmvrmc) [inline lib/services/runtime/types.ts:18]: arent we missing `DTO.isTextRequest` here? otherwise this will make an infinite loop

Comment (zhihil) [reply lib/services/runtime/types.ts:18]: You're right.

File: lib/services/runtime/types.ts [hunk 5]
```diff
@@ -13,34 +14,26 @@ export interface Prompt {
   voice?: string;
 }
 
-export const isTextRequest = (request?: RuntimeRequest | null): request is BaseRequest.TextRequest =>
-  !!request && BaseRequest.isTextRequest(request) && typeof request.payload === 'string';
+export const isTextRequest = (request?: RuntimeRequest): request is BaseRequest.TextRequest =>
+  !!request && DTO.isTextRequest(request);
```
Comment (zhihil) [inline lib/services/runtime/types.ts:18]: The `typeof request.payload === 'string'` check is redundant.

File: lib/services/slots/index.ts [hunk 6]
```diff
@@ -1,19 +1,28 @@
...
+      BaseRequest.isIntentRequest(value) &&
+      !!value.payload?.intent?.name &&
+      Array.isArray(value.payload.entities) &&
+      !value.payload?.data
+    );
+  }
+
   handle = async (context: Context) => {
-    if (!isIntentRequest(context.request)) {
+    if (!this.isLUISIntentRequest(context.request)) {
```
Comment (zhihil) [inline lib/services/slots/index.ts:25]: This is legacy (possibly unused) code which expects a value that doesn't define the required property `value` on slots. 

I don't think we should make `value` optional to accommodate this old type, so I've instead created a legacy type guard to allow the code to continue working.

File: lib/controllers/interact.ts [hunk 7]
```diff
@@ -31,6 +32,13 @@ class InteractController extends AbstractController {
       { locale?: string; logs: RuntimeLogs.LogLevel }
     >
   ): Promise<ResponseContext> {
+    if (req.body.request) {
+      req.body.request = AnyRequestDTO.parse(req.body.request);
+    }
+    if (req.body.action) {
+      req.body.action = AnyRequestDTO.parse(req.body.action);
+    }
```
Comment (zhihil) [inline lib/controllers/interact.ts:40]: Various controller methods have been updated to add Zod parsing to the request objects. This ensures that the incoming request data is valid and adapted when necessary, e.g, intent request objects may lack the required `entities` and `query` properties and the DTO provides default values for them.

File: lib/services/runtime/types.ts [hunk 8]
```diff
@@ -1,51 +1,61 @@
...
+export type PathRequest = z.infer<typeof PathRequestDTO>;
+
 export interface Prompt {
   content: BaseText.SlateTextValue | string;
   voice?: string;
 }
 
-export const isTextRequest = (request?: RuntimeRequest | null): request is BaseRequest.TextRequest =>
-  !!request && BaseRequest.isTextRequest(request) && typeof request.payload === 'string';
-
```
Comment (zhihil) [inline lib/services/runtime/types.ts:18]: Many of these type-guards have been deleted and replaced with type-guards from the DTO package, which have more stringent validation requirements.

File: lib/services/aiAssist.ts [hunk 9]
```diff
@@ -16,16 +17,16 @@ class AIAssist extends AbstractManager implements ContextHandler {
   static StringStorageKey = 'vf_memory';
 
   static getInput(request: RuntimeRequest) {
-    if (isIntentRequest(request)) {
+    if (DTO.isLegacyIntentRequest(request)) {
```
Comment (trs) [inline lib/services/aiAssist.ts:20]: Curious about the reason to have these under `DTO` rather than imported directly.

**Review by trs — APPROVED** [review]
This could cause some unintended issues with peoples requests, we'll definitely want to keep an eye on requests.

File: lib/controllers/interact/index.ts [hunk 10]
```diff
@@ -26,7 +27,7 @@ class InteractController extends AbstractController {
     const { versionID } = params;
     const { sessionID } = body.session;
     const { userID } = body.session;
-    const { action } = body;
+    const action = await AnyRequestDTO.parseAsync(req.body.action);
```
Comment (trs) [inline lib/controllers/interact/index.ts:30]: Could update the above `InteractRequestBody` to have the `AnyRequestDTO`, and not have to do parsing here.

**Review by trs — CHANGES_REQUESTED** [review]

File: lib/services/aiAssist.ts (hunk 9 above)
Comment (zhihil) [reply lib/services/aiAssist.ts:20]: It was to disambiguate it from the type guards that are in `general-runtime`. It's a style that's also used on `creator-app` to provide clearer separation between what Voiceflow code belongs to what package.

Is it fine to adopt this convention for DTOs here?

**Review by trs — APPROVED** [review]

**Pull request closed without merging**